src/remote.c

```diff
@@ -211,58 +211,79 @@ static int ensure_remote_doesnot_exist(git_repository *repo, const char *name)
+int git_remote_create_init_options(git_remote_create_options *opts, unsigned int version)
+{
+	GIT_INIT_STRUCTURE_FROM_TEMPLATE(
+		opts, version, git_remote_create_options, GIT_REMOTE_CREATE_OPTIONS_INIT);
+	return 0;
+}
+
+static int create_internal(git_remote **out, const char *url, const git_remote_create_options *opts)
 {
 	git_remote *remote;
 	git_config *config_ro = NULL, *config_rw;
 	git_buf canonical_url = GIT_BUF_INIT;
 	git_buf var = GIT_BUF_INIT;
+	const git_remote_create_options dummy_opts = GIT_REMOTE_CREATE_OPTIONS_INIT;
 	int error = -1;
+	int is_anonymous;
 
-	/* repo, name, and fetch are optional */
 	assert(out && url);
 
-	if (repo && (error = git_repository_config_snapshot(&config_ro, repo)) < 0)
+	if (!opts) {
+		opts = &dummy_opts;
+	}
+
+	is_anonymous = (!opts->name || !opts->repository);
+	if (!is_anonymous) {
+		if ((error = ensure_remote_name_is_valid(opts->name)) < 0)
+			return error;
+
+		if ((error = ensure_remote_doesnot_exist(opts->repository, opts->name)) < 0)
+			return error;
+	}
+
+	if (opts->repository &&
+		(error = git_repository_config_snapshot(&config_ro, opts->repository)) < 0)
 		return error;
 
 	remote = git__calloc(1, sizeof(git_remote));
 	GITERR_CHECK_ALLOC(remote);
 
-	remote->repo = repo;
+	remote->repo = opts->repository;
 
 	if ((error = git_vector_init(&remote->refs, 32, NULL)) < 0 ||
 		(error = canonicalize_url(&canonical_url, url)) < 0)
 		goto on_error;
 
-	if (repo) {
+	if (opts->repository) {
 		remote->url = apply_insteadof(config_ro, canonical_url.ptr, GIT_DIRECTION_FETCH);
 	} else {
 		remote->url = git__strdup(canonical_url.ptr);
 	}
 	GITERR_CHECK_ALLOC(remote->url);
 
-	if (name != NULL) {
-		remote->name = git__strdup(name);
+	if (opts->name != NULL) {
+		remote->name = git__strdup(opts->name);
 		GITERR_CHECK_ALLOC(remote->name);
 
-		if ((error = git_buf_printf(&var, CONFIG_URL_FMT, name)) < 0)
+		if ((error = git_buf_printf(&var, CONFIG_URL_FMT, opts->name)) < 0)
 			goto on_error;
 
-		if (repo &&
-			((error = git_repository_config__weakptr(&config_rw, repo)) < 0 ||
+		if (opts->repository &&
+			((error = git_repository_config__weakptr(&config_rw, opts->repository)) < 0 ||
 			(error = git_config_set_string(config_rw, var.ptr, canonical_url.ptr)) < 0))
 			goto on_error;
 	}
 
-	if (fetch != NULL) {
-		if ((error = add_refspec(remote, fetch, true)) < 0)
+	if (opts->fetchspec != NULL) {
+		if ((error = add_refspec(remote, opts->fetchspec, true)) < 0)
 			goto on_error;
 
 		/* only write for non-anonymous remotes */
-		if (repo && name && (error = write_add_refspec(repo, name, fetch, true)) < 0)
-			goto on_error;
-
-		if (repo && (error = lookup_remote_prune_config(remote, config_ro, name)) < 0)
+		if (!is_anonymous &&
+			((error = write_add_refspec(opts->repository, opts->name, opts->fetchspec, true)) < 0 ||
+			(error = lookup_remote_prune_config(remote, config_ro, opts->name)) < 0))
```

Comment:
Wait, this is different from what we did before. Now we only call `lookup_remote_prune_config` if both `opts->repository` and `opts->name` are set. Previously, we were also calling it in case `opts->name` is unset. Is this intentional and, if so, why can we avoid calling `lookup_remote_prune_config` now?

Reply:
Yeah, the tests around detached/anonymous/realized remotes are confusing here. My `is_anonymous` variable should be named `is_not_realized` maybe (bikeshedding welcome on the name).

The original code would attempt to load a `remote..prune` config option, which IMHO doesn't make sense (and I remember a test broke before I changed it to this).

Reply:
Confirmed. Should I keep the test as is ?
```
  1) Failure:
network::remote::remotes::cannot_add_a_nameless_remote [/Users/tiennou/Projects/libgit2/tests/network/remote/remotes.c:321]
  GIT_EINVALIDSPEC != git_remote_create(&remote, _repo, NULL, "git://github.com/libgit2/libgit2")
  -12 != 0
```

Reply:
You're right, the original code doesn't make a lot of sense. So this change looks correct to me

Reply:
Note that most of what I said above about the broken test is wrong. The test failure was caused when I removed the name & repo existence checks from `git_remote_create` (since it allows `NULL`).

AFAICT, the original code is extra-confusing about what actually needs a repo/name because it indirectly uses the config, so I hope this is better. I'm still dubious of the whole config_ro/config_rw dance that's happening, because we start by doing a snapshot, then we re-grab a "live" config. I might be just paranoid/unaware of the semantics of `config__weakptr`, but I'm tempted to change it to : grab a "live" version IF we have a repo, snapshot it so it doesn't change when applying `url.insteadof`, and set the various other settings on the first live one.